text/0000-async-project-goal-2024h2.md

```diff
@@ -0,0 +1,275 @@
+
+* async closures
+* a ["send bound"][sb] solution, most likely [RTN][]
+
+# Frequently asked questions
+
+## Why focus on send bounds + async closures?
+
+These are the two features that together block the authoring of traits for a number of common interop purposes. Send bounds are needed for generic traits like the `Service` trait. Async closures are needed for rich combinator APIs like iterators.
+
+## Why not work on dyn dispatch for async fn in traits?
+
+Async fn in traits do not currently support native dynamic dispatch. We have explored a [number of designs for making it work](https://smallcultfollowing.com/babysteps/blog/2021/09/30/dyn-async-traits-part-1/) but are not currently prioritizing that effort. It was determined that this idea is lower priority because it is possible to [workaround](https://smallcultfollowing.com/babysteps/blog/2021/10/15/dyn-async-traits-part-6/) the gap by having the  `#[trait_variant]` crate produce a dynamic dispatch wrapper type (e.g., `#[trait_variant(dyn = DynWidget)] trait Widget` would create a type `DynWidget<'_>` that acts like a `Box<dyn Widget>`). We do expect to support dyn async trait, hopefully in 2025.
+
+## Why are we moving forward on a trait for async iteration?
+
+There has been extensive discussion about the best design for the "Stream" or "async iter" trait and we judge that the design space is well understood. We would like to unblock generator syntax in 2025 which will require some form of trait.
+
+The majority of the debate about the trait has been on the topic of whether to base the trait on a `poll_next` function, as we do today, or to try and make the trait use `async fn next`, making it more anaologous with the `Iterator` trait (and potentially even making it be two versions of a single trait). We will definitely explore forwards compatibility questions as part of this discussion. nikomatsakis for example still wants to explore maybe-async-like designs, especially for combinator APIs like `map`. However, we also refer to the design axiom that that "when in doubt, zero-cost is our compass" -- we believe we should be able to stabilize a trait that does the low-level details right, and then design higher level APIs atop that.
+
+## Why work on a revised async vision doc? Don't we already have one?
+
+The existing doc was authored some time back and is in need of an update. Moreover, the original doc was never RFC'd and we have found that it lacks a certain measure of "authority" as a result. We would like to drive stronger alignment on the path forward so that we can focus more on execution.
+
+## What about "maybe async", effect systems, and keyword generics?
+
+Keyword generics is an ambitious initiative to enable code that is "maybe async". It has generated significant controversy, with some people feeling it is necessary for Rust to scale and others judging it to be overly complex. We anticipate having more debate on this topic as part of drafting the async vision doc.
```

Comment:
I don't want to set an expectation that we will settle this question fully.. but I think we can put some parameters around it.

Reply:
Update.